### PR TITLE
CVSL-1708 add soft delete audit events

### PR DESCRIPTION
**Pull request by louisenorris**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppshdcapi/licences/AuditEventRepository.kt [hunk 1]
```diff
@@ -13,4 +13,10 @@ interface AuditEventRepository : JpaRepository<AuditEvent, Long>, JpaSpecificati
     value = "select id, timestamp, user, action, details from audit where details ->> 'bookingId' in :bookingIds order by timestamp desc;",
   )
   fun findByBookingIds(bookingIds: Collection<String>): List<AuditEvent>
+
+  @Query(
+    nativeQuery = true,
+    value = "insert into audit user, action, details;",
+  )
```
Comment (louisenorris) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppshdcapi/licences/AuditEventRepository.kt:20]: Not sure if I also insert a `timestamp` in this query or if this gets added and populated automatically like I think `id` does? 😅

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppshdcapi/licences/softdelete/SoftDeleteService.kt [hunk 2]
```diff
@@ -62,7 +65,7 @@ class SoftDeleteService(
     val licencesRecords = licencesToMigrate(initialIdToProcess, numberToMigrate)
     val lastIdProcessed = licencesRecords.content.lastOrNull()?.first?.id
     log.info("Last Id processed in batch: ${lastIdProcessed ?: " no records processed"}")
-    val licencesToSoftDelete = applyAnySoftDeletes(licencesRecords)
+    val licencesToSoftDelete = applyAnySoftDeletes(licencesRecords, "MIGRATION")
```
Comment (SebNegahbanMoJ) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppshdcapi/licences/softdelete/SoftDeleteService.kt:68]: I'm not the best Kotlin-er, but I believe Andy has said to me previously that if there's a list of constant string values, it's best to put them in an enum and reference the enum.
eg CVL [AuditEventTypes](https://github.com/ministryofjustice/create-and-vary-a-licence-api/blob/main/src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/util/AuditEventType.kt) which is referenced like [this](https://github.com/ministryofjustice/create-and-vary-a-licence-api/blob/main/src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/jobs/DeactivateLicencesService.kt#L64)

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppshdcapi/integration/ResetLicenceTest.kt [hunk 3]
```diff
@@ -53,9 +53,9 @@ class ResetLicenceTest : SqsIntegrationTestBase() {
     assertThat(events).hasSize(3)
 
     assertThat(events).extracting("user", "action", "details").containsExactly(
-      tuple("SYSTEM:API", "RESET", mapOf("bookingId" to 10)),
```
Comment (louisenorris) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppshdcapi/integration/ResetLicenceTest.kt:56]: `:` is not a legal character in a Java identifier so have changed this and other event user types to use `_` as a word separator instead. Hope changing the user type from `SYSTEM:API` (a type I think has been saved to the db already) to `SYSTEM_API` doesn't have negative consequences in the db? Seemed wrong to not make the API type an enum also.

Comment (andrewrlee) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/hmppshdcapi/integration/ResetLicenceTest.kt:56]: you can have a text value associated with the enum value.

Comment (andrewrlee) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/hmppshdcapi/integration/ResetLicenceTest.kt:56]: https://kotlinlang.org/docs/enum-classes.html

```kt
enum class Color(val colourName: String) {
    RED("Red"),
    GREEN("Green"),
    BLUE("Blue")
}
```
so you can then refer to the string label as:
`Color.RED.colourName`

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppshdcapi/integration/SoftDeleteJobTest.kt [hunk 4]
```diff
@@ -69,6 +74,14 @@ class SoftDeleteJobTest : SqsIntegrationTestBase() {
     // We don't re-delete previously deleted licences
     assertThat(versions[16]).isEqualTo(LocalDateTime.of(2022, 7, 27, 15, 0, 0, 0))
     assertThat(versions[11]!!.toLocalDate()).isEqualTo(LocalDate.now())
+
+    val events = auditEventRepository.findAll()
+    assertThat(events).hasSize(2)
+
+    assertThat(events).extracting("user", "action", "details").containsExactly(
+      Assertions.tuple("SYSTEM_JOB", "RESET", mapOf("bookingId" to 10)),
```
Comment (andrewrlee) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppshdcapi/integration/SoftDeleteJobTest.kt:82]: **quibble:** would usually statically import things like this so you don't need to add `Assertions.`

**Review by gurps-singh — APPROVED** [review]

**Pull request merged**